IECoreGL::TextureLoader : Use OIIO and apply color convert correctly

danieldresser-ie opened the pull request:
Avoid use of deprecated IECoreImage::ImageReader and ToGLTextureConverter, and let OIIO do the color convert of the whole image at once, which means it will handle unpremulting and repremulting correctly.

When paired with https://github.com/GafferHQ/gaffer/pull/4269, this will result in perfect edges on Gaffer icons.  Without that PR, it will remove the dark edges, but not fix the overexpanded alpha edges, resulting in icons looking slightly too thick.

Code looks pretty good to me, but I still need to do some testing of images with data windows, and using maximumResolution to set the mip level.

danieldresser-ie commented:
Hmm, bunch of annoying test failures here.  The Linux python2 failure is spurious.  The windows one is pretty annoying - looks like importing OIIO brought in a Windows header processenv.h which has a `#define` that redefines "SearchPath" to "SearchPathA"

And the most annoying is the ImageEngine ones - I was using up to date OIIO calls, which aren't supported by the old OIIO that IE is using.

johnhaddon commented on src/IECoreGL/TextureLoader.cpp:
```diff
@@ -69,49 +75,108 @@ TexturePtr TextureLoader::load( const std::string &name, int maximumResolution )
+
+	TexturePtr t = nullptr;
+	if ( !y && r && g && b )
+    {
+        t = new ColorTexture( imageBuf.spec().full_width, imageBuf.spec().full_height, r.get(), g.get(), b.get(), a.get() );
+    }
```
Tabs not spaces

johnhaddon commented on src/IECoreGL/TextureLoader.cpp:
```diff
@@ -69,49 +75,108 @@ TexturePtr TextureLoader::load( const std::string &name, int maximumResolution )
+        t = new ColorTexture( imageBuf.spec().full_width, imageBuf.spec().full_height, r.get(), g.get(), b.get(), a.get() );
+    }
+	else if ( y && !r && !g && !b )
+    {
+        t = new LuminanceTexture( imageBuf.spec().full_width, imageBuf.spec().full_height, y.get(), a.get() );
+    }
```
Tabs not spaces

johnhaddon reviewed (changes requested):
> processenv.h which has a #define that redefines "SearchPath" to "SearchPathA"

Looks like we deal with this via `#undef SearchPath` in `IECore/SearchPath.h`. Probably just need the same in `TextureLoader.cpp`?

> I was using up to date OIIO calls, which aren't supported by the old OIIO that IE is using.

Oh, I thought the latest IE environment was pretty much up to date now? Are we planning on this change going into `10.3` or `10.2`? If the former, then perhaps IE could bump up OIIO at the same time?

andrewkaufman commented:
> I thought the latest IE environment was pretty much up to date now?

We're using OIIO 2.2.6.1 (released Sep 1, 2020), so its not all that modern anymore I suppose. That's still the same major version as GafferHQ though so I'm a bit surprised they'd have broken API compatibility...

> perhaps IE could bump up OIIO for 10.3

I'd hope to do that (along with moving to gcc 9, c++17, boost 1.73, etc) but it will most likely depend on the timing of 10.3 as we're not ready for all that internally (or within Gaffer) yet...

danieldresser-ie commented:
OK, I think this is good to go.  Included SearchPath.h again to get the hack that fixes the define, updated the dependencies version to get the tests passing and fixed the formatting issues you noted.

To get the mipmap tests passing, I'm now creating a temp ImageCache with automip turned on - it seems fine to throw out the ImageCache each time when we're caching the GL texture ourselves?  It would probably be better to only use a mipmap if one exists on disk, and otherwise if we're oversize just explicitly resize to maximumResolution, but I'm matching the old behaviour for now.

Anyway, that gets the mipmap test passing, and I've also added a test for dataWindows not matching the display window, so I think this should be good to go.

johnhaddon commented:
LGTM.

> I was using up to date OIIO calls, which aren't supported by the old OIIO that IE is using.

I think this turned out to be the outdated CI dependencies you fixed in 4df3b76, correct? So IE _is_ using a sufficiently new version of OIIO? If so, maybe we want to merge this to `RB-10.2`?

@andrewkaufman, we'll definitely want this in for Gaffer 0.61, and I'm really hoping for a smaller gap between major versions this time. Is that another argument for it going in `RB-10.2`, or are you expecting to be able to deploy `10.3` soon enough for 0.61?

andrewkaufman commented:
I would hope that something this small doesn't merit 10.3 all on its own. Is there even a compatibility break here? IECoreGL already required OIIO via IECoreImage, so its just a visual change which we're claiming as a bug fix, right?

If there is still some OIIO version issue, I'd hope that can be handled with conditional compilation rather than raising the minimum required OIIO.

johnhaddon commented:
> Is there even a compatibility break here? IECoreGL already required OIIO via IECoreImage, so its just a visual change which we're claiming as a bug fix, right?

That's my understanding, yep. I'm all for it going into `RB-10.2` but I just want to check on the OIIO version compatibility at IE before merging (I think it's fine even without conditional compilation, but just want confirmation).

danieldresser-ie commented:
> > 
> 
> I think this turned out to be the outdated CI dependencies you fixed in [4df3b76](https://github.com/ImageEngine/cortex/commit/4df3b76b8b29c2843aa9d1f2033e0a30663df86a), correct? So IE _is_ using a sufficiently new version of OIIO? If so, maybe we want to merge this to `RB-10.2`?
> 
> @

Oh, yeah, I should have explicitly confirmed that.  Yeah, IE is fine, it was just the CI tests named "ImageEngine.cortex" that threw me off, and those are fixed by 4df3b76.  I would say it's fine to go in RB-10.2

johnhaddon commented:
Cool - could you rebase this PR onto `RB-10.2` then Daniel, and I'll merge?

danieldresser-ie commented:
Rebased onto RB-10.2
